astropy/units/core.py

```diff
@@ -1959,6 +1959,10 @@ def _unrecognized_operator(self, *args, **kwargs):
     __pow__ = __truediv__ = __rtruediv__ = __mul__ = __rmul__ = _unrecognized_operator
     __lt__ = __gt__ = __le__ = __ge__ = __neg__ = _unrecognized_operator
 
+    def __hash__(self):
+        # __hash__ isn't inherited in classes with a custom __eq__ method
+        return self._hash
```

Comment:
really not sure this is worth adding a test for, but I'll leave it to reviewers to advise.